trustymail/trustymail.py

```diff
@@ -63,6 +68,76 @@ def mx_scan(domain):
         handle_error("[MX]", domain, error)
 
 
+def starttls_scan(domain, timeout):
+    """
+    Scan a domain to see if it sends mail and supports STARTTLS.
+
+    Scan a domain to see if it sends mail.  If the domain does send
+    mail, a further check will be done to see if it supports STARTTLS.
+    All results are stored inside the Domain object that is passed in
+    as a parameter.
+
+    Parameters
+    ----------
+    domain : Domain
+        The Domain to be tested.
+        
+    timeout : int
+        The SMTP connection timeout in seconds.  (Default is 5.)
+    """
+    for mail_server in domain.mail_servers:
+        for port in SMTP_PORTS:
+            smtp_connection = smtplib.SMTP(timeout=timeout)
+            server_and_port = mail_server + ":" + str(port)
+            logging.debug("Testing " + server_and_port + " for STARTTLS support")
+            # Try to connect.  This will tell us if something is
+            # listening.
+            try:
+                smtp_connection.connect(mail_server, port)
+            except (socket.timeout, smtplib.SMTPConnectError, smtplib.SMTPServerDisconnected, ConnectionRefusedError, OSError) as error:
+                handle_error("[STARTTLS]", domain, error)
+                tmp = server_and_port + "-" + str(False)
+                domain.sends_mail.append(tmp)
+                domain.starttls.append(tmp)
+                continue
+
+            # Now try to say hello.  This will tell us if the thing
+            # that is listening is an SMTP server.
+            try:
+                smtp_connection.ehlo_or_helo_if_needed()
```

Comment:
The docs for [`ehlo_or_helo_if_needed()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.ehlo_or_helo_if_needed) indicate it calls [`ehlo()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.ehlo)/[`helo()`](https://docs.python.org/3/library/smtplib.html#smtplib.SMTP.helo), which has `name` as a parameter. This parameter "defaults to the fully qualified domain name of the local host", which, in our and an arbitrary user's case, seems liable to be an .in-addr.arpa address or just the IP.

Per the 6th and 7th paragraphs of [section 4.1.4. of RFC 5321](https://tools.ietf.org/html/rfc5321#section-4.1.4), the current SMTP standard, this looks to be ok (if an address doesn't reverse to be an FQDN, SMTP will use the raw address). Perhaps we should run a few tests and see if hardcoding (or making configurable) an EHLO name (postmaster@.gov) makes any difference in our results?

Reply:
I will add a flag for setting the name.  By default it will work as it does now, but if you specify the name it will be used instead.